File: src/Queue/CircularQueue.java
package Queue;

public class CircularQueue<T> implements QueueInterface<T> {
	
	private T[] array;
	private int tail;
	private int head;
	private int elements;
	
	public CircularQueue(int size) {
		this.array = (T[]) new Object[size];
		this.tail = -1;
		this.head = -1;
		this.elements = 0;
	}
	
	@Override
	public void enqueue(T element) throws QueueOverflowException {
		if(this.isFull()){
			throw new QueueOverflowException();
		}
		if(element != null){
			array[++tail % this.array.length] = element;
			elements++;
		}
	}

	@Override
	public T dequeue() throws QueueUnderflowException {
		if(this.isEmpty()){
			throw new QueueUnderflowException();
		}
		T element = array[++head % this.array.length];
		this.elements--;
		return element;
	}

	@Override
	public T head() {
		if(isEmpty())
			return null;
		else
			return this.array[(head + 1) % this.array.length];
	}

	@Override
	public boolean isEmpty() {
		return this.elements == 0;
	}

	@Override
	public boolean isFull() {
		return this.elements == this.array.length;
	}

}

File: Python/Queue/CircularQueue.py
'''
	Author: Ionésio Junior
'''
#coding:utf -8
'''
	Circular Queue Implementation
'''
class CircularQueue():
	
	def __init__(self,size = 10):
		self.__queueArray = []
		self.__tail = -1
		self.__head = -1
		self.__elements = 0
		self.__size = size
	
	
	def enqueue(self,element):
		if(self.isFull()):
			raise Exception("Queue is Full!!")
		else:
			if(element != None):
				self.__elements = self.__elements + 1
				if(self.__head == -1 and self.__tail == -1):
					self.__head = (self.__head + 1) % self.__size
				
				self.__tail = (self.__tail + 1) % self.__size
				self.__queueArray.insert(self.__tail,element)	
			
				
	def dequeue(self):
		if(self.isEmpty()):
			raise Exception("Queue is Empty!!")
		else:
			self.__elements = self.__elements - 1
			removedElement = self.__queueArray[self.__head]
			self.__head = (self.__head + 1) % self.__size
			return removedElement
		

	def head(self):
		if(self.isEmpty()):
			return None
		else:
			return self.__queueArray[self.__head]
	
	
	def isEmpty(self):
		return (self.__elements == 0)
	
	
	def isFull(self):
		return (self.__elements == self.__size)

